bindings/rust/s2n-tls/src/renegotiate.rs

```diff
@@ -59,21 +176,111 @@ impl RenegotiateCallback for RenegotiateResponse {
-    ///    ([Connection::set_receive_callback()], [Connection::set_receive_context()])
+    ///    [Connection::set_receive_callback()], [Connection::set_receive_context()]
+    ///
+    /// In addition, the Rust bindings do not wipe:
+    /// - The server name: [Connection::set_server_name()]. The s2n-tls-tokio
+    ///   TlsConnector sets the server name automatically, so preserving it across
+    ///   wipes prevents all users of s2n-tls-tokio from needing a custom callback
+    ///   just to maintain consistent behavior.
+    /// - The waker: [Connection::set_waker()]. Wiping the waker during a call
+    ///   to `poll_send` or `poll_recv` can break IO.
+    ///
+    /// The set of configuration values that are not wiped may change in the future.
+    /// Therefore if you specifically need certain connection configuration values
+    /// wiped during renegotiation, then you should wipe them yourself in
+    /// [RenegotiateCallback::on_renegotiate_wipe()].
     pub fn wipe_for_renegotiate(&mut self) -> Result<(), Error> {
-        self.wipe_method(|conn| unsafe { s2n_renegotiate_wipe(conn.as_ptr()).into_result() })
+        // Check for buffered data in order to surface more specific
+        // error messages to the application.
+        if self.renegotiate_state().send_pending {
+            return Err(Error::from(
+                ErrorType::UsageError,
+                "RenegotiateError",
+                "Unexpected buffered send data during renegotiate",
+            ));
+        }
+
+        // Save any state that needs to be preserved.
+        // The only real cost of saving state here is complexity. We can't save all
+        // connection configuration automatically because in the C library, connection
+        // configuration is indistinguishable from C connection state.
+        let renegotiate_state = *self.renegotiate_state();
+        let waker = self.waker().map(|waker| waker.clone());
+        let server_name = self.server_name().map(|name| name.to_owned());
+
+        self.wipe_method(|conn| unsafe { s2n_renegotiate_wipe(conn.as_ptr()).into_result() })?;
+
+        // Restore the saved state
+        *self.renegotiate_state_mut() = renegotiate_state;
+        self.set_waker(waker.as_ref())?;
+        if let Some(server_name) = server_name {
+            self.set_server_name(&server_name)?;
+        }
+
+        // We trigger the callback last so that the application can modify any
+        // preserved configuration (like the server name or waker) if necessary.
+        if let Some(mut config) = self.config() {
+            if let Some(callback) = config.context_mut().renegotiate.as_mut() {
+                callback.on_renegotiate_wipe(self)?;
+            }
+        }
+
+        self.renegotiate_state_mut().needs_wipe = false;
+        Ok(())
+    }
+
+    fn poll_renegotiate_raw(
+        &mut self,
+        buf_ptr: *mut libc::c_void,
+        buf_len: isize,
+    ) -> (Poll<Result<(), Error>>, usize) {
```

Comment:
The problem is that that "usize" is unrelated to the actual completion of the poll_renegotiate. It's just something that can happen along the way. If I switch to Poll<Result<usize, Error>>, the caller will need to keep polling after Ready, because Ready does not mean finished. There aren't any real rules for general poll_* methods, but there are rules for Future::poll, and that violates the "Once a future has finished, clients should not poll it again" one. Needing to say "renegotiate is only done once you get Ready(Ok(0)) instead of Ready(Ok(x>0))" seems worse than the awkward but explicit signature :/